File: src/main/java/org/example/kalkulationsprogramm/controller/KalenderController.java
package org.example.kalkulationsprogramm.controller;

import lombok.RequiredArgsConstructor;
import org.example.kalkulationsprogramm.domain.KalenderEintrag;
import org.example.kalkulationsprogramm.domain.Mitarbeiter;
import org.example.kalkulationsprogramm.repository.MitarbeiterRepository;
import org.example.kalkulationsprogramm.service.KalenderService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/kalender")
@RequiredArgsConstructor
public class KalenderController {

    private final KalenderService kalenderService;
    private final MitarbeiterRepository mitarbeiterRepository;

    /**
     * DTO für Teilnehmer.
     */
    public record TeilnehmerDto(Long id, String name) {
        static TeilnehmerDto fromEntity(Mitarbeiter m) {
            return new TeilnehmerDto(m.getId(), m.getVorname() + " " + m.getNachname());
        }
    }

    /**
     * DTO für Kalendereinträge (zur Vermeidung von Lazy-Loading-Problemen).
     */
    public record KalenderEintragDto(
            Long id,
            String titel,
            String beschreibung,
            LocalDate datum,
            LocalTime startZeit,
            LocalTime endeZeit,
            boolean ganztaegig,
            String farbe,
            Long projektId,
            String projektName,
            Long kundeId,
            String kundeName,
            Long lieferantId,
            String lieferantName,
            Long anfrageId,
            String anfrageBetreff,
            Long erstellerId,
            String erstellerName,
            List<TeilnehmerDto> teilnehmer) {

        static KalenderEintragDto fromEntity(KalenderEintrag e) {
            List<TeilnehmerDto> teilnehmerDtos = new ArrayList<>();
            if (e.getTeilnehmer() != null) {
                teilnehmerDtos = e.getTeilnehmer().stream()
                        .map(TeilnehmerDto::fromEntity)
                        .collect(Collectors.toList());
            }

            return new KalenderEintragDto(
                    e.getId(),
                    e.getTitel(),
                    e.getBeschreibung(),
                    e.getDatum(),
                    e.getStartZeit(),
                    e.getEndeZeit(),
                    e.isGanztaegig(),
                    e.getFarbe(),
                    e.getProjekt() != null ? e.getProjekt().getId() : null,
                    e.getProjekt() != null ? e.getProjekt().getBauvorhaben() : null,
                    e.getKunde() != null ? e.getKunde().getId() : null,
                    e.getKunde() != null ? e.getKunde().getName() : null,
                    e.getLieferant() != null ? e.getLieferant().getId() : null,
                    e.getLieferant() != null ? e.getLieferant().getLieferantenname() : null,
                    e.getAnfrage() != null ? e.getAnfrage().getId() : null,
                    e.getAnfrage() != null ? e.getAnfrage().getBauvorhaben() : null,
                    e.getErsteller() != null ? e.getErsteller().getId() : null,
                    e.getErsteller() != null ? e.getErsteller().getVorname() + " " + e.getErsteller().getNachname()
                            : null,
                    teilnehmerDtos);
        }
    }

    /**
     * Lädt alle Einträge für einen Monat.
     * Optional: Wenn mitarbeiterId angegeben, werden nur relevante Einträge für
     * diesen Mitarbeiter gezeigt.
     */
    @GetMapping
    public ResponseEntity<List<KalenderEintragDto>> getEintraege(
            @RequestParam int jahr,
            @RequestParam int monat,
            @RequestParam(required = false) Long mitarbeiterId) {
        List<KalenderEintrag> eintraege;
        if (mitarbeiterId != null) {
            // Gefilterte Ansicht: Nur Termine für diesen Mitarbeiter
            eintraege = kalenderService.getEintraegeForMitarbeiter(mitarbeiterId, jahr, monat);
        } else {
            // Keine Filterung: Alle Termine (für Admin-Ansicht etc.)
            eintraege = kalenderService.getEintraegeForMonat(jahr, monat);
        }
        List<KalenderEintragDto> dtos = eintraege.stream()
                .map(KalenderEintragDto::fromEntity)
                .collect(Collectors.toList());
        return ResponseEntity.ok(dtos);
    }

    // ==================== MOBILE ENDPOINTS (Token-basiert) ====================

    /**
     * Lädt alle Einträge für einen Mitarbeiter in einem Monat (Mobile App).
     */
    @GetMapping("/mobile")
    public ResponseEntity<List<KalenderEintragDto>> getEintraegeMobile(
            @RequestParam String token,
            @RequestParam int jahr,
            @RequestParam int monat) {
        Mitarbeiter mitarbeiter = mitarbeiterRepository.findByLoginToken(token).orElse(null);
        if (mitarbeiter == null || !Boolean.TRUE.equals(mitarbeiter.getAktiv())) {
            return ResponseEntity.status(401).build();
        }

        List<KalenderEintrag> eintraege = kalenderService.getEintraegeForMitarbeiter(mitarbeiter.getId(), jahr, monat);
        List<KalenderEintragDto> dtos = eintraege.stream()
                .map(KalenderEintragDto::fromEntity)
                .collect(Collectors.toList());
        return ResponseEntity.ok(dtos);
    }

    /**
     * Lädt alle Einträge für einen Mitarbeiter an einem Tag (Tagesansicht).
     */
    @GetMapping("/mobile/tag")
    public ResponseEntity<List<KalenderEintragDto>> getEintraegeForTag(
            @RequestParam String token,
            @RequestParam String datum) {
        Mitarbeiter mitarbeiter = mitarbeiterRepository.findByLoginToken(token).orElse(null);
        if (mitarbeiter == null || !Boolean.TRUE.equals(mitarbeiter.getAktiv())) {
            return ResponseEntity.status(401).build();
        }

        LocalDate localDate = LocalDate.parse(datum);
        List<KalenderEintrag> eintraege = kalenderService.getEintraegeForMitarbeiterTag(mitarbeiter.getId(), localDate);
        List<KalenderEintragDto> dtos = eintraege.stream()
                .map(KalenderEintragDto::fromEntity)
                .collect(Collectors.toList());
        return ResponseEntity.ok(dtos);
    }

    /**
     * Lädt einen einzelnen Eintrag.
     */
    @GetMapping("/{id}")
    public ResponseEntity<KalenderEintragDto> getEintrag(@PathVariable Long id) {
        KalenderEintrag eintrag = kalenderService.getEintragWithTeilnehmer(id);
        if (eintrag == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(KalenderEintragDto.fromEntity(eintrag));
    }

    /**
     * Request-Body für Erstellen/Aktualisieren.
     */
    public record KalenderEintragRequest(
            String titel,
            String beschreibung,
            LocalDate datum,
            LocalTime startZeit,
            LocalTime endeZeit,
            boolean ganztaegig,
            String farbe,
            Long projektId,
            Long kundeId,
            Long lieferantId,
            Long anfrageId,
            List<Long> teilnehmerIds,
            Long erstellerId) {
    }

    /**
     * Erstellt einen neuen Kalendereintrag (PC-Frontend).
     */
    @PostMapping
    public ResponseEntity<KalenderEintragDto> createEintrag(@RequestBody KalenderEintragRequest request) {
        KalenderEintrag eintrag = new KalenderEintrag();
        eintrag.setTitel(request.titel());
        eintrag.setBeschreibung(request.beschreibung());
        eintrag.setDatum(request.datum());
        eintrag.setStartZeit(request.startZeit());
        eintrag.setEndeZeit(request.endeZeit());
        eintrag.setGanztaegig(request.ganztaegig());
        eintrag.setFarbe(request.farbe());

        KalenderEintrag saved = kalenderService.saveEintrag(
                eintrag,
                request.projektId(),
                request.kundeId(),
                request.lieferantId(),
                request.anfrageId(),
                request.erstellerId(),
                request.teilnehmerIds());
        return ResponseEntity.ok(KalenderEintragDto.fromEntity(saved));
    }

    /**
     * Erstellt einen neuen Kalendereintrag (Mobile App).
     */
    @PostMapping("/mobile")
    public ResponseEntity<KalenderEintragDto> createEintragMobile(
            @RequestParam String token,
            @RequestBody KalenderEintragRequest request) {
        Mitarbeiter mitarbeiter = mitarbeiterRepository.findByLoginToken(token).orElse(null);
        if (mitarbeiter == null || !Boolean.TRUE.equals(mitarbeiter.getAktiv())) {
            return ResponseEntity.status(401).build();
        }

        KalenderEintrag eintrag = new KalenderEintrag();
        eintrag.setTitel(request.titel());
        eintrag.setBeschreibung(request.beschreibung());
        eintrag.setDatum(request.datum());
        eintrag.setStartZeit(request.startZeit());
        eintrag.setEndeZeit(request.endeZeit());
        eintrag.setGanztaegig(request.ganztaegig());
        eintrag.setFarbe(request.farbe());

        KalenderEintrag saved = kalenderService.saveEintrag(
                eintrag,
                request.projektId(),
                request.kundeId(),
                request.lieferantId(),
                request.anfrageId(),
                mitarbeiter.getId(),
                request.teilnehmerIds());
        return ResponseEntity.ok(KalenderEintragDto.fromEntity(saved));
    }

    /**
     * Aktualisiert einen bestehenden Kalendereintrag.
     */
    @PutMapping("/{id}")
    public ResponseEntity<KalenderEintragDto> updateEintrag(
            @PathVariable Long id,
            @RequestBody KalenderEintragRequest request) {
        KalenderEintrag existing = kalenderService.getEintrag(id);
        if (existing == null) {
            return ResponseEntity.notFound().build();
        }

        existing.setTitel(request.titel());
        existing.setBeschreibung(request.beschreibung());
        existing.setDatum(request.datum());
        existing.setStartZeit(request.startZeit());
        existing.setEndeZeit(request.endeZeit());
        existing.setGanztaegig(request.ganztaegig());
        existing.setFarbe(request.farbe());

        KalenderEintrag saved = kalenderService.saveEintrag(
                existing,
                request.projektId(),
                request.kundeId(),
                request.lieferantId(),
                request.anfrageId(),
                null,
                request.teilnehmerIds());
        return ResponseEntity.ok(KalenderEintragDto.fromEntity(saved));
    }

    /**
     * Löscht einen Kalendereintrag.
     */
    @DeleteMapping("/{id}")
    public ResponseEntity<Map<String, Boolean>> deleteEintrag(@PathVariable Long id) {
        kalenderService.deleteEintrag(id);
        return ResponseEntity.ok(Map.of("deleted", true));
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/controller/KalenderController.kt
package org.example.kalkulationsprogramm.controller

import org.example.kalkulationsprogramm.domain.KalenderEintrag
import org.example.kalkulationsprogramm.domain.Mitarbeiter
import org.example.kalkulationsprogramm.repository.MitarbeiterRepository
import org.example.kalkulationsprogramm.service.KalenderService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.time.LocalDate
import java.time.LocalTime

@RestController
@RequestMapping("/api/kalender")
class KalenderController(
    private val kalenderService: KalenderService,
    private val mitarbeiterRepository: MitarbeiterRepository,
) {
    data class TeilnehmerDto(
        val id: Long?,
        val name: String,
    ) {
        companion object {
            fun fromEntity(m: Mitarbeiter): TeilnehmerDto =
                TeilnehmerDto(m.id, "${m.vorname} ${m.nachname}")
        }
    }

    data class KalenderEintragDto(
        val id: Long?,
        val titel: String?,
        val beschreibung: String?,
        val datum: LocalDate?,
        val startZeit: LocalTime?,
        val endeZeit: LocalTime?,
        val ganztaegig: Boolean,
        val farbe: String?,
        val projektId: Long?,
        val projektName: String?,
        val kundeId: Long?,
        val kundeName: String?,
        val lieferantId: Long?,
        val lieferantName: String?,
        val anfrageId: Long?,
        val anfrageBetreff: String?,
        val erstellerId: Long?,
        val erstellerName: String?,
        val teilnehmer: List<TeilnehmerDto>,
    ) {
        companion object {
            fun fromEntity(e: KalenderEintrag): KalenderEintragDto {
                val teilnehmerDtos = e.teilnehmer.map(TeilnehmerDto::fromEntity)
                val projekt = e.projekt
                val kunde = e.kunde
                val lieferant = e.lieferant
                val anfrage = e.anfrage
                val ersteller = e.ersteller

                return KalenderEintragDto(
                    id = e.id,
                    titel = e.titel,
                    beschreibung = e.beschreibung,
                    datum = e.datum,
                    startZeit = e.startZeit,
                    endeZeit = e.endeZeit,
                    ganztaegig = e.isGanztaegig(),
                    farbe = e.farbe,
                    projektId = projekt?.id,
                    projektName = projekt?.bauvorhaben,
                    kundeId = kunde?.id,
                    kundeName = kunde?.name,
                    lieferantId = lieferant?.id,
                    lieferantName = lieferant?.lieferantenname,
                    anfrageId = anfrage?.id,
                    anfrageBetreff = anfrage?.bauvorhaben,
                    erstellerId = ersteller?.id,
                    erstellerName = ersteller?.let { "${it.vorname} ${it.nachname}" },
                    teilnehmer = teilnehmerDtos,
                )
            }
        }
    }

    @GetMapping
    fun getEintraege(
        @RequestParam jahr: Int,
        @RequestParam monat: Int,
        @RequestParam(required = false) mitarbeiterId: Long?,
    ): ResponseEntity<List<KalenderEintragDto>> {
        val eintraege =
            if (mitarbeiterId != null) {
                kalenderService.getEintraegeForMitarbeiter(mitarbeiterId, jahr, monat)
            } else {
                kalenderService.getEintraegeForMonat(jahr, monat)
            }
        return ResponseEntity.ok(eintraege.map(KalenderEintragDto::fromEntity))
    }

    @GetMapping("/mobile")
    fun getEintraegeMobile(
        @RequestParam token: String,
        @RequestParam jahr: Int,
        @RequestParam monat: Int,
    ): ResponseEntity<List<KalenderEintragDto>> {
        val mitarbeiter = mitarbeiterRepository.findByLoginToken(token).orElse(null)
        if (mitarbeiter == null || mitarbeiter.aktiv != true) {
            return ResponseEntity.status(401).build()
        }

        val eintraege = kalenderService.getEintraegeForMitarbeiter(mitarbeiter.id, jahr, monat)
        return ResponseEntity.ok(eintraege.map(KalenderEintragDto::fromEntity))
    }

    @GetMapping("/mobile/tag")
    fun getEintraegeForTag(
        @RequestParam token: String,
        @RequestParam datum: String,
    ): ResponseEntity<List<KalenderEintragDto>> {
        val mitarbeiter = mitarbeiterRepository.findByLoginToken(token).orElse(null)
        if (mitarbeiter == null || mitarbeiter.aktiv != true) {
            return ResponseEntity.status(401).build()
        }

        val localDate = LocalDate.parse(datum)
        val eintraege = kalenderService.getEintraegeForMitarbeiterTag(mitarbeiter.id, localDate)
        return ResponseEntity.ok(eintraege.map(KalenderEintragDto::fromEntity))
    }

    @GetMapping("/{id}")
    fun getEintrag(@PathVariable id: Long): ResponseEntity<KalenderEintragDto> {
        val eintrag = kalenderService.getEintragWithTeilnehmer(id)
        return ResponseEntity.ok(KalenderEintragDto.fromEntity(eintrag))
    }

    data class KalenderEintragRequest(
        val titel: String?,
        val beschreibung: String?,
        val datum: LocalDate?,
        val startZeit: LocalTime?,
        val endeZeit: LocalTime?,
        val ganztaegig: Boolean,
        val farbe: String?,
        val projektId: Long?,
        val kundeId: Long?,
        val lieferantId: Long?,
        val anfrageId: Long?,
        val teilnehmerIds: List<Long>?,
        val erstellerId: Long?,
    )

    @PostMapping
    fun createEintrag(@RequestBody request: KalenderEintragRequest): ResponseEntity<KalenderEintragDto> {
        val eintrag = KalenderEintrag().apply {
            titel = request.titel
            beschreibung = request.beschreibung
            datum = request.datum
            startZeit = request.startZeit
            endeZeit = request.endeZeit
            ganztaegig = request.ganztaegig
            farbe = request.farbe
        }

        val saved = kalenderService.saveEintrag(
            eintrag,
            request.projektId,
            request.kundeId,
            request.lieferantId,
            request.anfrageId,
            request.erstellerId,
            request.teilnehmerIds,
        )
        return ResponseEntity.ok(KalenderEintragDto.fromEntity(saved))
    }

    @PostMapping("/mobile")
    fun createEintragMobile(
        @RequestParam token: String,
        @RequestBody request: KalenderEintragRequest,
    ): ResponseEntity<KalenderEintragDto> {
        val mitarbeiter = mitarbeiterRepository.findByLoginToken(token).orElse(null)
        if (mitarbeiter == null || mitarbeiter.aktiv != true) {
            return ResponseEntity.status(401).build()
        }

        val eintrag = KalenderEintrag().apply {
            titel = request.titel
            beschreibung = request.beschreibung
            datum = request.datum
            startZeit = request.startZeit
            endeZeit = request.endeZeit
            ganztaegig = request.ganztaegig
            farbe = request.farbe
        }

        val saved = kalenderService.saveEintrag(
            eintrag,
            request.projektId,
            request.kundeId,
            request.lieferantId,
            request.anfrageId,
            mitarbeiter.id,
            request.teilnehmerIds,
        )
        return ResponseEntity.ok(KalenderEintragDto.fromEntity(saved))
    }

    @PutMapping("/{id}")
    fun updateEintrag(
        @PathVariable id: Long,
        @RequestBody request: KalenderEintragRequest,
    ): ResponseEntity<KalenderEintragDto> {
        val existing = kalenderService.getEintrag(id) ?: return ResponseEntity.notFound().build()

        existing.apply {
            titel = request.titel
            beschreibung = request.beschreibung
            datum = request.datum
            startZeit = request.startZeit
            endeZeit = request.endeZeit
            ganztaegig = request.ganztaegig
            farbe = request.farbe
        }

        val saved = kalenderService.saveEintrag(
            existing,
            request.projektId,
            request.kundeId,
            request.lieferantId,
            request.anfrageId,
            null,
            request.teilnehmerIds,
        )
        return ResponseEntity.ok(KalenderEintragDto.fromEntity(saved))
    }

    @DeleteMapping("/{id}")
    fun deleteEintrag(@PathVariable id: Long): ResponseEntity<Map<String, Boolean>> {
        kalenderService.deleteEintrag(id)
        return ResponseEntity.ok(mapOf("deleted" to true))
    }
}
